src/bitrot.py

```diff
@@ -345,36 +361,32 @@ def report_done(
         if self.test and self.verbosity:
             print('warning: database file not updated on disk (test mode).')
 
-    def handle_unknown_path(self, cur, new_path, new_mtime, new_sha1):
+    def handle_unknown_path(self, cur, new_path, new_mtime, new_sha1, paths_fs, hashes_dict_db):
         """Either add a new entry to the database or update the existing entry
         on rename.
 
         Returns `new_path` if the entry was indeed new or the `stored_path` (e.g.
         outdated path) if there was a rename.
         """
-        cur.execute('SELECT mtime, path, timestamp FROM bitrot WHERE hash=?',
-                    (new_sha1,))
-        rows = cur.fetchall()
-        for row in rows:
-            stored_mtime, stored_path, stored_ts = row
-            if os.path.exists(stored_path):
-                # file still exists, move on
-                continue
 
-            # update the path in the database
+        try: # if doesn't exist the fs path of the database
+            found = [path for path in hashes_dict_db[new_sha1] if path not in paths_fs]
+            renamed = found.pop()
+            if renamed :
+                # update the path in the database
+                cur.execute(
+                    'UPDATE bitrot SET mtime=?, path=?, timestamp=? WHERE path=?',
+                    (new_mtime, new_path, ts(), renamed),
+                )
+
+                return renamed
+        except:
```

Comment:
The exception was for hashes_dict_db[new_sha1], as new_sha1 could be a new one sha1 indeed. ;)

I'm sure this block can be reworked in something simpler/cleaner.